src/structures/interfaces/Collector.js

```diff
@@ -147,6 +147,38 @@ class Collector extends EventEmitter {
     });
   }
 
+  /**
+   * Allows this collector to be consumed with `for await...of`
+   * @type {AsyncIterator}
+   */
+  async * [Symbol.asyncIterator]() {
+    const queue = [];
+    const onCollect = item => {
+      queue.push(item);
+    };
+
+    this.on('collect', onCollect);
+
+    while (!this.ended || queue.length > 0) {
+      if (queue.length > 0) {
+        yield queue.shift();
+      } else {
+        await new Promise(resolve => { // eslint-disable-line no-await-in-loop
+          const tick = () => {
+            this.removeListener('collect', tick);
+            this.removeListener('end', tick);
+            resolve();
+          };
+
+          this.on('collect', tick);
+          this.on('end', tick);
+        });
+      }
+    }
+
+    this.removeListener('collect', onCollect);
```

Comment:
This should be in a `finally` branch because the `yield` statement on like 164 can return early if a consumer calls `return` or `throw` on the generator object.

Reply:
Yes, please add a try {} finally {} @patrickroberts

Reply:
@bdistin Should `this.on('collect', onCollect);` be inside or outside the `try` block?

Reply:
Outside, it never errors. This line should be moved inside a finally block, as @bdistin noted.

Reply:
try { the while loop } finally { remove the listener}... Everything else as normal